test/Test/Database/LSMTree/StateMachine.hs

```diff
@@ -2005,35 +1997,17 @@ arbitraryActionWithVars _ label ctx (ModelState st _stats) =
         , isJust (eqT @h @ModelIO.Table)
         ]
       where
-        -- Generate at least a 2-way union, and at most a 3-way union.
+        -- For querying the union debt or supplying union credits, the
+        -- interesting cases to test for are when tables are union tables. For
+        -- non-union tables, these operations should just be no-ops, so we
+        -- generate them only rarely.
         --
-        -- Unit tests for 0-way and 1-way unions are included in the UnitTests
-        -- module. n-way unions for n>3 lead to larger unions, which are less likely
-        -- to be finished before the end of an action sequence.
-        gen2or3TableVars :: Gen (NonEmpty (Var h (WrapTable h IO k v b)))
-        gen2or3TableVars = do
-            tableVar1 <- genTableVar
-            tableVar2 <- genTableVar
-            mtableVar3 <- QC.oneof [pure Nothing, Just <$> genTableVar]
-            pure $ NE.fromList $ catMaybes [
-                Just tableVar1, Just tableVar2, mtableVar3
-              ]
-
         -- TODO: tweak distribution once table unions are implemented
-        genUnionTableVar = QC.frequency $
-            -- The interesting cases to test for are when tables are union
-            -- tables.
-            [ (9, genUnionDescendantTableVars)
-            | not (null unionDescendantTableVars)
-            ]
-            -- For non-union tables, querying the union debt or supplying union
-            -- credits are no-ops, so we generate such tables only rarely.
-            --
-            -- TODO: replace union actions on non-union tables with a few unit
-            -- tests?
-         ++ [ (1, genNotUnionDescendantTableVars)
-            | not (null notUnionDescendantTableVars)
-            ]
+        -- TODO: replace union actions on non-union tables with unit tests?
+        genUnionTableVar = QC.frequency [
+            (9 * length unionDescendantTableVars,    genUnionDescendantTableVar)
+          , (1 * length notUnionDescendantTableVars, genNotUnionDescendantTableVar)
+          ]
```

Comment:
So the idea is to skew more towards union descendants?

Reply:
Yes, though in a follow-up patch I made this even more extreme and moved the trivial cases of non-union tables to a unit test, and replaced `genUnionTableVar` with `genUnionDescendantTableVar` so the QLS test only covers the non-trivial cases of union-derived tables.

Reply:
BTW, the original code does not do what it intends to do. It intends to have 90% : 10% skew toward union tables, but in fact the result is 40% : 60% skew, i.e. 60% are trivial non-union tables.

I fix this in a later patch (by removing the trivial case entirely), but it could also be fixed by moving the two cases up into the parent `QC.frequency`, each with their corresponding guard. That way we don't get forced into picking the trivial case when there are no non-trivial cases available. It's very common that there are no union tables available.

Reply:
:+1:  but even in the parent `QC.frequency` you'd get a different distribution than the integers you supply, depending on what actions are available and which guards are true/false. But I agree it's still better

Reply:
Yes you would, but that should then work as intended.